src/main/java/org/candlepin/subscriptions/tally/UsageSnapshotProducer.java

```diff
@@ -84,28 +85,36 @@ public UsageSnapshotProducer(FactNormalizer factNormalizer, AccountListSource ac
     public void produceSnapshots() {
         try {
             List<String> accountList = accountListSource.list();
-            log.info("Producing snapshots for {} accounts in batches of {}.", accountList.size(),
-                accountBatchSize);
-
-            // Partition the account list to help reduce memory usage while performing the calculations.
-            int count = 0;
-            for (List<String> accounts : Iterables.partition(accountList, accountBatchSize)) {
-                Collection<AccountUsageCalculation> accountCalcs =
-                    accountUsageCollector.collect(APPLICABLE_PRODUCTS, accounts);
-                dailyRoller.rollSnapshots(accounts, accountCalcs);
-                weeklyRoller.rollSnapshots(accounts, accountCalcs);
-                monthlyRoller.rollSnapshots(accounts, accountCalcs);
-                yearlyRoller.rollSnapshots(accounts, accountCalcs);
-                quarterlyRoller.rollSnapshots(accounts, accountCalcs);
-                count += accounts.size();
-                log.info("{}/{} accounts processed.", count, accountList.size());
-            }
-            log.info("Finished producing snapshots for all accounts.");
+            produceSnapshotsForAccounts(accountList);
```

Comment:
See [pitfall 2](https://codete.com/blog/5-common-spring-transactional-pitfalls/)

```

* The method visibility can’t be any other than public.
* The invocation must come from outside of the bean.
```

So the internal invocation here will not be transactional; we need to refactor out the call to another class. :slightly_frowning_face:

Reply:
@awood, thanks for the info; I believe it's okay in this case because `produceSnapshotsForAccounts(...)` is a private method, and the two public methods:
 - `produceSnapshotsForAccount`
 - `produceSnapshots`

are both annotated with `@Transactional`.

Reply:
i.e. all entrypoints to public methods are transactional, and there are no call paths that involve trying to invoke a transactional method from inside the bean.

Reply:
Ah, the diff cut off that the method you changed was already transactional.